src/agent/coverage/src/code.rs

```diff
@@ -0,0 +1,331 @@
+                // need to compute its file offset.
+                //
+                // A symbol is defined relative to some section, identified by `st_shndx`, an index
+                // into the section header table. We'll use the section header to compute the file
+                // offset of the symbol.
+                let section = object
+                    .section_headers
+                    .get(sym.st_shndx)
+                    .cloned()
+                    .ok_or_else(|| format_err!("invalid section table index for symbol"))?;
+
+                // If mapped into a segment, `sh_addr` contains the VA of the section image,
+                // consistent with the `p_vaddr` of the segment.
+                //
+                // https://refspecs.linuxbase.org/elf/gabi4+/ch4.sheader.html#section_header
+                let section_va = section.sh_addr;
+
+                // The offset of the symbol relative to its section (both in-file and when mapped).
+                let sym_section_offset = sym_va - section_va;
+
+                // We have the file offset for the section via `sh_offset`, and the offset of the
+                // symbol within the section. From this, calculate the file offset for the symbol,
+                // which we can use to index into `data`.
+                let sym_file_offset = section.sh_offset + sym_section_offset;
+
+                let entry = Symbol {
+                    name,
+                    file_offset: sym_file_offset,
+                    image_offset,
+                    size: sym.st_size,
+                };
+
+                let inserted = symbols.index.insert(entry.clone());
+                if !inserted {
+                    log::error!("failed to insert symbol index entry: {:x?}", entry);
+                }
+            }
+        }
+
+        Ok(Self {
+            path,
+            base_va,
+            symbols,
+        })
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(transparent)]
+pub struct SymbolIndex {
+    pub index: RegionIndex<Symbol>,
+}
+
+impl SymbolIndex {
+    pub fn iter(&self) -> impl Iterator<Item = &Symbol> {
+        self.index.iter()
+    }
+
+    /// Find the symbol metadata for the image-relative `offset`.
+    pub fn find(&self, offset: u64) -> Option<&Symbol> {
+        self.index.find(offset)
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Symbol {
+    /// Raw symbol name, possibly mangled.
+    pub name: String,
+
+    /// File offset of the symbol definition in the on-disk module.
+    pub file_offset: u64,
+
+    /// Module-relative offset of the mapped symbol.
+    pub image_offset: u64,
+
+    /// Total size in bytes of the symbol definition.
+    pub size: u64,
+}
+
+impl Symbol {
+    pub fn contains_file_offset(&self, offset: u64) -> bool {
+        let lo = self.file_offset;
+        let hi = lo + self.size;
```

Comment:
Should this be a checked add?

Reply:
If the add were to overflow, the `Symbol` info itself is nonsense. This could happen with a malformed binary file, so we should really validate earlier.